server/src/com/cloud/storage/snapshot/SnapshotManagerImpl.java

```diff
@@ -854,15 +855,17 @@ public SnapshotPolicyVO createPolicy(CreateSnapshotPolicyCmd cmd, Account policy
                 throw new InvalidParameterValueException("Max number of snapshots shouldn't exceed the " + message + " level snapshot limit");
             }
         }
-        SnapshotPolicyVO policy = _snapshotPolicyDao.findOneByVolumeInterval(volumeId, intvType);
-        if (policy == null) {
-            policy = new SnapshotPolicyVO(volumeId, cmd.getSchedule(), timezoneId, intvType, cmd.getMaxSnaps(), display);
-            policy = _snapshotPolicyDao.persist(policy);
-            _snapSchedMgr.scheduleNextSnapshotJob(policy);
-        } else {
-            try {
+
+        policy = _snapshotPolicyDao.acquireInLockTable(volumeId);
```

Comment:
You can request a lock for the snapshot policy table for existing entries and by giving a snapshot policy ID only, not the ID of a volume: `policy = _snapshotPolicyDao.acquireInLockTable(policyId);`
In your case you don't have a policy id yet when creating a new one. You'll have to use the global lock.

Reply:
@marcaurele : Changed the code as per your suggestion. Now added a code for using global lock along with volume id added in the lock name.
Can you please review the new code ?